File: apps/api/app/schemas/graphs.py
"""Graph schemas."""

from typing import Any

from pydantic import BaseModel, Field


class GraphNode(BaseModel):
    id: str
    label: str
    type: str  # 'paper' | 'author' | 'topic'
    degree: int = 0
    cluster_id: int | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class GraphEdge(BaseModel):
    source: str
    target: str
    type: str  # 'cites' | 'co_authored' | 'shared_topic' | 'similar_embedding'
    weight: float = 1.0


class TopicCluster(BaseModel):
    cluster_id: int
    label: str
    paper_count: int
    keywords: list[str]


class ResearchGraphResponse(BaseModel):
    project_id: str
    nodes: list[GraphNode]
    edges: list[GraphEdge]
    total_papers: int
    total_authors: int
    total_topics: int
    clusters: list[TopicCluster]
    bridge_papers: list[str]


class DiscoveryRecommendation(BaseModel):
    id: str
    title: str
    authors: list[str]
    year: int | None = None
    abstract: str | None = None
    doi: str | None = None
    arxiv_id: str | None = None
    relevance_score: float | None = None
    reason: str
    source_topics: list[str]

File: apps/web/src/lib/api/graphs.ts
import { request } from './client';

export interface GraphNodeDTO {
  id: string;
  label: string;
  type: 'paper' | 'author' | 'topic' | string;
  paper_id?: string;
  author_name?: string;
  topic_name?: string;
  degree?: number;
  cluster_id?: number;
  year?: number;
  citation_count?: number;
  metadata?: Record<string, any>;
  x?: number;
  y?: number;
}

export interface GraphEdgeDTO {
  source: string;
  target: string;
  relationship?: string;
  type?: string;
  weight?: number;
}

export interface GraphClusterDTO {
  id?: number;
  cluster_id?: number;
  label: string;
  size?: number;
  paper_count?: number;
  color?: string;
  keywords?: string[];
}

export interface ResearchGraphDTO {
  project_id: string;
  nodes: GraphNodeDTO[];
  edges: GraphEdgeDTO[];
  total_papers: number;
  total_authors: number;
  total_topics: number;
  clusters: GraphClusterDTO[];
  bridge_papers: string[];
}

export interface RelatedPaperDTO {
  id: string;
  title: string;
  authors: string[];
  year?: number;
  abstract?: string;
  similarity_score: number;
  relevance_score?: number | null;
  reason?: string;
  doi?: string;
  arxiv_id?: string;
  pmid?: string;
  citation_count?: number;
}

export const graphsApi = {
  getResearchGraph: (projectId: string) =>
    request<ResearchGraphDTO>(`/projects/${projectId}/research-graph`),
  discoverRelated: (projectId: string) =>
    request<RelatedPaperDTO[]>(`/projects/${projectId}/discover-related`),
};
